<a href="https://colab.research.google.com/github/obengdouglas/DSI-2022/blob/main/DSI_imdb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Pre-DSI Assignment - 2022**

**Question 2: Datascience with Pandas and Movie Data**

**Notes**

Pandas, jupyter notebooks and colab are some of the most basic python tools in data science. Here we want to make sure you know how to load data into pandas, manipulate columns, rows and subsets of dataframes, perform string manipulation, regex, merging of dataframes, groupby, sorting of values, as well as know how to perform basic statistics on dataframes.

We also want you to be familiar with installing python packages, reading documentation and have a little familiarity with key libraries like scipy, sklearn and pytorch. 

Finally we also want you to show that you understand the basics of git, which is an industry standard. 


[url to document](https://docs.google.com/document/d/1ZUltBJgPx2d3Y3QC44xoibZJFCpNo2mIR9KZkpnLdv8/edit)

In [ ]:
#import libraries

import numpy as np # linear algebra
import pandas as pd # data processing
import os
import matplotlib.pyplot as plt
import numpy as np
from functools import reduce

import seaborn as sns
import sklearn
import scipy
import torch
# from sklearn import preprocessing, svm
# from sklearn.model_selection import train_test_split
# from sklearn.linear_model import LinearRegression

In [ ]:
#import files
#0. Access some historical IMDB data files from the shared drive

from google.colab import drive
drive.mount("/content/drive")

my_path = "/Colab_Files/DSI_imdb" 
gdrive_path = "/content/drive" + "/My Drive" + my_path

#sample_input_file = gdrive_path + "input.csv" # The specific file you are trying to access
#rawdata = pd.read_csv(sample_input_file)
# /content/drive/My Drive/folders/my_folder_name/input.csv

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
#1. Read the files 'title.basics.tsv.gz', 'title.akas.tsv.gz' and 'title.ratings.tsv.gz' into three separate dataframes using the read_csv method in Pandas.

akas      = pd.read_csv(gdrive_path + '/title.akas.tsv.gz', compression='gzip', sep='\t', quotechar='"', error_bad_lines=False)
basics    = pd.read_csv(gdrive_path + '/title.basics.tsv.gz', compression='gzip', sep='\t', quotechar='"', error_bad_lines=False)
reviews   = pd.read_csv(gdrive_path + '/title.ratings.tsv.gz', compression='gzip', sep='\t', quotechar='"', error_bad_lines=False)

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (7) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)
/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (5) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [ ]:
akas.describe()
akas.head()

titleId  ordering  ... attributes isOriginalTitle
0  tt0000001         1  ...         \N               0
1  tt0000001         2  ...         \N               0
2  tt0000001         3  ...         \N               0
3  tt0000001         4  ...         \N               1
4  tt0000002         1  ...         \N               1

[5 rows x 8 columns]

In [ ]:
akas_2 = akas.rename(columns={'titleId': 'tconst'})
akas_2.head(10)


tconst  ordering  ...             attributes isOriginalTitle
0  tt0000001         1  ...                     \N               0
1  tt0000001         2  ...                     \N               0
2  tt0000001         3  ...                     \N               0
3  tt0000001         4  ...                     \N               1
4  tt0000002         1  ...                     \N               1
5  tt0000002         2  ...                     \N               0
6  tt0000002         3  ...                     \N               0
7  tt0000002         4  ...                     \N               0
8  tt0000002         5  ...  literal English title               0
9  tt0000002         6  ...                     \N               0

[10 rows x 8 columns]

In [ ]:
basics.head(10)

tconst titleType  ... runtimeMinutes                    genres
0  tt0000001     short  ...              1         Documentary,Short
1  tt0000002     short  ...              5           Animation,Short
2  tt0000003     short  ...              4  Animation,Comedy,Romance
3  tt0000004     short  ...             \N           Animation,Short
4  tt0000005     short  ...              1              Comedy,Short
5  tt0000006     short  ...              1                     Short
6  tt0000007     short  ...              1               Short,Sport
7  tt0000008     short  ...              1         Documentary,Short
8  tt0000009     movie  ...             45                   Romance
9  tt0000010     short  ...              1         Documentary,Short

[10 rows x 9 columns]

In [ ]:
reviews.head(10)

tconst  averageRating  numVotes
0  tt0000001            5.8      1422
1  tt0000002            6.4       168
2  tt0000003            6.6      1015
3  tt0000004            6.4       100
4  tt0000005            6.2      1713
5  tt0000006            5.5        88
6  tt0000007            5.5       573
7  tt0000008            5.6      1526
8  tt0000009            5.5        70
9  tt0000010            6.9      5083

In [ ]:
#2. Drop duplicates in all the dataframes, if there are any. 

akas_2.drop_duplicates()
basics.drop_duplicates()
reviews.drop_duplicates()


tconst  averageRating  numVotes
0       tt0000001            5.8      1422
1       tt0000002            6.4       168
2       tt0000003            6.6      1015
3       tt0000004            6.4       100
4       tt0000005            6.2      1713
...           ...            ...       ...
874198  tt9100972            7.0         5
874199  tt9104810            7.0         7
874200  tt9106766            7.6         5
874201  tt9107292            8.4        10
874202  tt9112162            9.9        10

[874203 rows x 3 columns]

In [ ]:
#3. Using the Pandas 'merge' method, combine all three dataframes using the Title ID (titleID or tconst) to perform the merge and save it into a new dataframe. 

imdb_list = [akas_2, basics, reviews]
imdb_merge = reduce(lambda left,right: pd.merge(left,right,on='tconst'), imdb_list)
imdb_merge

tconst  ordering  ... averageRating numVotes
0        tt0000001         1  ...           5.8     1422
1        tt0000001         2  ...           5.8     1422
2        tt0000001         3  ...           5.8     1422
3        tt0000001         4  ...           5.8     1422
4        tt0000002         1  ...           6.4      168
...            ...       ...  ...           ...      ...
1686542  tt9100068         2  ...           7.0        5
1686543  tt9100098         1  ...           7.0        5
1686544  tt9100098         2  ...           7.0        5
1686545  tt9100972         1  ...           7.0        5
1686546  tt9100972         2  ...           7.0        5

[1686547 rows x 18 columns]

In [ ]:
imdb_outer = reduce(lambda left,right: pd.merge(left,right, how='outer', on='tconst'), imdb_list)
imdb_outer

tconst  ordering  ... averageRating numVotes
0        tt0000001       1.0  ...           5.8   1422.0
1        tt0000001       2.0  ...           5.8   1422.0
2        tt0000001       3.0  ...           5.8   1422.0
3        tt0000001       4.0  ...           5.8   1422.0
4        tt0000002       1.0  ...           6.4    168.0
...            ...       ...  ...           ...      ...
6740084  tt9120652       NaN  ...           NaN      NaN
6740085  tt9120654       NaN  ...           NaN      NaN
6740086  tt9120656       NaN  ...           NaN      NaN
6740087  tt9120782       NaN  ...           NaN      NaN
6740088  tt9120906       NaN  ...           NaN      NaN

[6740089 rows x 18 columns]

In [ ]:
imdb_left = reduce(lambda left,right: pd.merge(left,right, how='left', on='tconst'), imdb_list)
imdb_left

tconst  ordering  ... averageRating numVotes
0        tt0000001         1  ...           5.8   1422.0
1        tt0000001         2  ...           5.8   1422.0
2        tt0000001         3  ...           5.8   1422.0
3        tt0000001         4  ...           5.8   1422.0
4        tt0000002         1  ...           6.4    168.0
...            ...       ...  ...           ...      ...
3423655  tt9118622         2  ...           NaN      NaN
3423656  tt9118622         3  ...           NaN      NaN
3423657  tt9118622         4  ...           NaN      NaN
3423658  tt9118622         5  ...           NaN      NaN
3423659  tt9118622         6  ...           NaN      NaN

[3423660 rows x 18 columns]

In [ ]:
imdb_right = reduce(lambda left,right: pd.merge(left,right, how='right', on='tconst'), imdb_list)
imdb_right

tconst  ordering  ... averageRating numVotes
0        tt0000001       1.0  ...           5.8     1422
1        tt0000001       2.0  ...           5.8     1422
2        tt0000001       3.0  ...           5.8     1422
3        tt0000001       4.0  ...           5.8     1422
4        tt0000002       1.0  ...           6.4      168
...            ...       ...  ...           ...      ...
1975951  tt9100972       2.0  ...           7.0        5
1975952  tt9104810       NaN  ...           7.0        7
1975953  tt9106766       NaN  ...           7.6        5
1975954  tt9107292       NaN  ...           8.4       10
1975955  tt9112162       NaN  ...           9.9       10

[1975956 rows x 18 columns]

In [ ]:
imdb_inner = reduce(lambda left,right: pd.merge(left,right, how='inner', on='tconst'), imdb_list)
imdb_inner

tconst  ordering  ... averageRating numVotes
0        tt0000001         1  ...           5.8     1422
1        tt0000001         2  ...           5.8     1422
2        tt0000001         3  ...           5.8     1422
3        tt0000001         4  ...           5.8     1422
4        tt0000002         1  ...           6.4      168
...            ...       ...  ...           ...      ...
1686542  tt9100068         2  ...           7.0        5
1686543  tt9100098         1  ...           7.0        5
1686544  tt9100098         2  ...           7.0        5
1686545  tt9100972         1  ...           7.0        5
1686546  tt9100972         2  ...           7.0        5

[1686547 rows x 18 columns]

**3.1 How many lines does the resulting dataframe have if you use an inner merge or outer  merge? Make sure you understand the difference.**

Answer

Number of lines with inner merge = 1686547 (default merge)

Number of lines with outer merge = 6740089


In [ ]:
imdb_inner.titleType.unique()


array(['short', 'movie', 'tvMovie', 'tvSeries', 'tvEpisode', 'tvShort',
       'tvMiniSeries', 'tvSpecial', 'video', 'videoGame'], dtype=object)

In [ ]:
# 3.2 Using the unique() method, compute how many different 'titleTypes' there are

imdb_inner.titleType.nunique()


10

In [ ]:
# 4. Make a new dataframe from step 3 by selecting only rows corresponding to English-language films ('en') OR US-region films ('US') 
# AND only those that are movies (using the 'titleType' column). Put the resulting data into a new dataframe; call it df_new.

df_new0 = imdb_inner[imdb_inner["language"].isin(['en', 'us'])]
df_new = df_new0[df_new0["titleType"].isin(['movie'])]
df_new

tconst  ordering  ... averageRating numVotes
1862     tt0000941         2  ...           4.9        7
2097     tt0001122         4  ...           5.9        7
2117     tt0001175         1  ...           6.0       21
2181     tt0001258         1  ...           5.7       62
2576     tt0001790         3  ...           6.0       20
...            ...       ...  ...           ...      ...
1686420  tt9020894         2  ...           4.5       20
1686485  tt9056646         3  ...           7.0        9
1686494  tt9060058         3  ...           8.1       17
1686520  tt9081472         2  ...           9.1      213
1686522  tt9081572         2  ...           6.6        8

[61365 rows x 18 columns]

In [ ]:
# 5. Add a new column to df_new with column title 'log10Votes' which gives the Log_10 number of the 'numVotes' column. 

df_new=df_new.copy()
df_new['log10Votes'] = np.log10(df_new['numVotes'])
df_new

tconst  ordering  ... numVotes log10Votes
1862     tt0000941         2  ...        7   0.845098
2097     tt0001122         4  ...        7   0.845098
2117     tt0001175         1  ...       21   1.322219
2181     tt0001258         1  ...       62   1.792392
2576     tt0001790         3  ...       20   1.301030
...            ...       ...  ...      ...        ...
1686420  tt9020894         2  ...       20   1.301030
1686485  tt9056646         3  ...        9   0.954243
1686494  tt9060058         3  ...       17   1.230449
1686520  tt9081472         2  ...      213   2.328380
1686522  tt9081572         2  ...        8   0.903090

[61365 rows x 19 columns]

In [ ]:
#6. Lower the case of all text in the 'genres' column.

df_new['genres']=df_new['genres'].str.lower()
df_new.head()

tconst  ordering  ... numVotes log10Votes
1862  tt0000941         2  ...        7   0.845098
2097  tt0001122         4  ...        7   0.845098
2117  tt0001175         1  ...       21   1.322219
2181  tt0001258         1  ...       62   1.792392
2576  tt0001790         3  ...       20   1.301030

[5 rows x 19 columns]

In [ ]:
# 6. Using Groupby (or other technique) group all data by 'genres' and display the top 10 highest genres by:
#     6.1 mean number of log10Votes
#     6.2 mean averageRating

df_new=df_new.groupby('genres')
df_new.first()

tconst  ordering  ... numVotes log10Votes
genres                                    ...                    
\n                   tt0001122         4  ...        7   0.845098
action               tt0034147         2  ...       18   1.255273
action,adult,comedy  tt0183256         1  ...       97   1.986772
action,adult,crime   tt0179274         1  ...       58   1.763428
action,adventure     tt0012484         1  ...       22   1.342423
...                        ...       ...  ...      ...        ...
thriller,war         tt0031968        19  ...     1579   3.198382
thriller,western     tt0044953         6  ...     8781   3.943544
war                  tt0033417         9  ...       16   1.204120
war,western          tt0045184        10  ...     1491   3.173478
western              tt0006375         2  ...       10   1.000000

[827 rows x 18 columns]

In [ ]:
#display the top 10 highest genres by:
#     6.1 mean number of log10Votes

df_new.agg({'log10Votes' : 'mean', 'averageRating' : 'mean'}).sort_values("log10Votes", ascending=False).head(10)

log10Votes  averageRating
genres                                               
adventure,sci-fi,thriller     5.803146       7.475000
mystery,romance,sci-fi        5.641197       7.500000
adventure,comedy,music        5.200092       6.400000
film-noir,mystery,thriller    5.138634       8.200000
film-noir,mystery             5.133127       8.100000
action,adventure,thriller     4.532179       6.407143
action,thriller,war           4.500103       6.885714
horror,musical,sci-fi         4.340484       6.700000
action,sci-fi,sport           4.296819       6.600000
biography,fantasy,horror      4.271842       5.400000

In [ ]:
#display the top 10 highest genres by:
#     6.2 mean averageRating

df_new.agg({'log10Votes' : 'mean', 'averageRating' : 'mean'}).sort_values("averageRating", ascending=False).head(10)

log10Votes  averageRating
genres                                                  
drama,history,news               0.698970           9.20
documentary,fantasy              1.050185           9.15
documentary,family,music         1.113943           8.90
animation,crime,documentary      1.431364           8.70
crime,documentary,thriller       1.593285           8.50
action,adventure,musical         1.633468           8.50
documentary,drama,thriller       1.135400           8.50
adventure,documentary,sport      2.283954           8.50
biography,documentary,musical    1.431364           8.40
biography,documentary,western    1.041393           8.40

In [ ]:
#7. Using ‘groupby’ group all data by averageRating and make a scatter plot of averageRating vs log10Votes.

df_new=df_new.groupby('averageRating')
df_new.first()

AttributeError: ignored

In [ ]:
#df_new_averageRating.plot.scatter(x = 'averageRating', y = 'log10Votes', s = 100);

fig, ax = plt.subplots(figsize=(12,12))
df_new.groupby('averageRating').plot(kind='scatter', x = "averageRating", y = "log10Votes", ax=ax)

In [ ]:
# 8. Perform linear regression on your data (averageRating vs log10Votes) created in the previous step in three different ways:

#     8.1 Using sklearn 
#     8.2 Using scipy 
#     8.3 Using pytorch 


